funcs_utils.php

```diff
@@ -0,0 +1,300 @@
+    }
+    $json = json_decode(file_get_contents($filename), true);
+    if (is_null($json)) {
+        $lastError = json_last_error();
+        error("Could not parse from $filename - last error was $lastError");
+    }
+    $modules = [];
+    foreach ($json as $module) {
+        $ghrepo = $module['github'];
+        $modules[] = [
+            'ghrepo' => $ghrepo,
+            'account' => explode('/', $ghrepo)[0],
+            'repo' => explode('/', $ghrepo)[1],
+            'cloneUrl' => "git@github.com:$ghrepo.git",
+            'branch' => max($module['branches'] ?: [-1])
+        ];
+    }
+    return $modules;
+}
+
+/**
+ * Returns a list of all scripts files to run against a particular cms major version
+ */
+function script_files($cmsMajor)
+{
+    if (!ctype_digit($cmsMajor)) {
+        $cmsMajor = "-$cmsMajor";
+    }
+    $scriptFiles = [];
+    $dir = "scripts/cms$cmsMajor";
+    if (!file_exists($dir)) {
+        warning("$dir does not exist, no CMS $cmsMajor specific scripts will be run");
+        return $scriptFiles;
+    }
+    if (!is_dir($dir)) {
+        error("$dir is not a directory");
+    }
+    if ($handle = opendir($dir)) {
+        while (false !== ($scriptFile = readdir($handle))) {
+            if ('.' === $scriptFile || '..' === $scriptFile) {
+                continue;
+            }
+            $scriptFiles[] = "$dir/$scriptFile";
+        }
+        closedir($handle);
+    }
+    return $scriptFiles;
+}
+
+/**
+ * Runs a shell command and returns the output
+ */
+function cmd($cmd, $cwd)
+{
+    info("Running command: $cmd in $cwd");
+    // using Process::fromShellCommandline() instead of new Process() so that pipes work
+    $process = Process::fromShellCommandline($cmd, $cwd);
+    $process->run();
+    if (!$process->isSuccessful()) {
+        warning("Error running command: $cmd in $cwd");
+        error("Output was: " . $process->getErrorOutput());
+    }
+    return trim($process->getOutput());
+}
+
+/**
+ * Returns a object used to output to the console
+ */
+function io(): SymfonyStyle
+{
+    global $IN, $OUT;
+    return new SymfonyStyle($IN ?: new ArgvInput(), $OUT ?: new NullOutput);
+}
+
+/**
+ * Removes a directory
+ */
+function remove_dir($dirname)
+{
+    if (!file_exists(($dirname))) {
+        return;
+    }
+    if (!is_dir($dirname)) {
+        error("$dirname is not a directory");
+    }
+    info("Removing $dirname");
+    shell_exec("rm -rf $dirname");
+}
+
+/**
+ * Validates the users system is ready to run the script
+ */
+function validate_system()
+{
+    $token = github_token();
+    if (!$token || !is_string($token)) {
+        error('Could not get github token - set MS_GITHUB_TOKEN environment variable');
+    }
+}
+
+/**
+ * Reads MS_GITHUB_TOKEN environment variable
+ */
+function github_token()
+{
+    return getenv('MS_GITHUB_TOKEN') ?: '';
+}
+
+/**
+ * Makes a request to the github API
+ */
+function github_api($url, $data = [])
+{
+    $token = github_token();
```

Comment:
Could we, please, check here as well, that `$token` is string.

Reply:
There is a check for is_string() below

Reply:
It's already validated during `validate_system()`

Reply:
But you get token here and then pass it to the `"Authorization: Bearer $token"` without any validation

Reply:
As I mentioned earlier, we've already validated that token in a non empty string in `validate_system()` earlier in this script.